src/middleware/createRouter.ts

```diff
@@ -0,0 +1,92 @@
+import simpleStateMapping from '../lib/stateMappings/simple';
+import historyRouter from '../lib/routers/history';
+import { Index } from '../widgets/index/index';
+import { Middleware } from '.';
+import { Router, StateMapping, UiState } from '../types';
+
+const walk = (current: Index, callback: (index: Index) => void) => {
+  callback(current);
+  current
+    .getWidgets()
+    .filter(function(widget): widget is Index {
+      return widget.$$type === 'ais.index';
+    })
+    .forEach(innerIndex => {
+      walk(innerIndex, callback);
+    });
+};
+
+interface RouterProps {
```

Comment:
The `Routing` type defined in `InstantSearch.ts` could be merged with this interface, no?

Reply:
Indeed.